tensorflow/core/kernels/maxpooling_op_sycl.h

```diff
@@ -111,7 +111,8 @@ struct LaunchMaxPoolingOpSYCL {
 
     const int output_size = output->NumElements();
     const int workgroup_size = device.maxSyclThreadsPerBlock();
-    const int n_threads = RoundUpToNearestMultiple(output_size, workgroup_size);
+    const int n_threads = output_size > workgroup_size ? RoundUpToNearestMultiple(output_size, workgroup_size) :
```

Comment:
the s no need for this check. RoundUpToNearestMultiple will do the check if it is not the multiple of 2 to round int to multiple of to. Otherwise if (index < output_size_)  inside the kernel will check the else part of this condition.

Reply:
The implementation of RoundUpToNearestMultiple needs to be fixed. Currently it will always round the value up, even if the output_size is already a multiple. It should be changed to something like (note the extra if...) 
```
template <typename IntegerType>
inline TF_ATTRIBUTE_ALWAYS_INLINE IntegerType
RoundUpToNearestMultiple(IntegerType val, const IntegerType multiplier) {
  static_assert(
      std::is_integral<IntegerType>::value,
      "Rounding to nearest multiple is only valid for integer types.");
  const IntegerType diff = val % multiplier;
  if (diff > 0) val += (multiplier - diff);
  return val;
}
```

Reply:
This condition is probably not useful for real application because the output_size should be bigger than the workgroup_size anyway. However the number of threads was 256 when my output size was 9 which is useless.
The implementation of RoundUpToNearestMultiple is at the beginning of the file and is not meant to check the output size. I think the cleanest solution would be to have a compute_kernel_range that computes the global size (and maybe the local size) of the kernel we want to launch but this is outside of this scope.

Reply:
After discussing with @mehdi-goli we think we could use the parallel_for_setup from Eigen which does this. Would that be ok?

Reply:
Eigen sycl device computes the compute_kernel_range inside parallel_for_setup(). So it should be able to use Eigen sycl device to set the parallel for parameters.

Reply:
Some thoughts:

1. I really don't think we should have branches or specialisations for small tensors which are not suited to running on GPUs anyway. This just creates additional maintenance work with no benefit to us.
2. Eigen's `parallel_for_setup` is more complex than we need here but would certainly do the job, however...
3. For code like this which we might want to release in a more standalone package at some point we want to be decoupling it from external libraries like Eigen.
4. The ideal solution is to query `CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE ` with `clGetKernelWorkGroupInfo`. If we can do this with SYCL, then with my code it is easy to see how to adapt the sizes to ensure that the number of threads launched is the correct multiple.

Personally I see no problem with launching 256 threads to do 9 work items worth of work. The GPU won't notice any difference and the performance difference will almost certainly be negligible.

Reply:
Ok fair points.
I still think it's important that we use the same function everywhere in TF for a simpler maintenance. Doesn't really matter if it is the Eigen one or a function based on RoundUpToNearestMultiple. We could copy this function in another package if need be.
Using CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE is a good idea we can query this in SYCL with the get_info method from a device.
 I guess I just like the idea of using as little resources as possible but it's true for GPU it doesn't always make sense.

This discussion should move in another JIRA, I'll just revert this as it was.